src/pages/blog/2024-06-28-why-i-like-graphql.mdx

```diff
@@ -0,0 +1,92 @@
+
+## Demand Control
+
+We cannot assume that all requests are equally hard on the server, whether we're using GraphQL or not.  The truth is that no matter what API style you use, whether that's a binary protocol over UDP or GraphQL, it is extremely rare, especially as the API surface grows, that all use-cases and "requests" will be equally expensive for a server to process.
+
+A very easy example to show this is simply a paginated endpoint:
+
+```
+GET /users?first=100
+
+GET /users?first=1
+```
+
+Or super expensive mutations:
+
+```
+POST /expensive-creation
+```
+
+To be 100% fair, of course GraphQL exposes this problem a bit more, earlier on, especially when not using persisted queries (Which should not happen!). And while folks building a small RPC API may not need to implement variable rate limiting or some sort of cost categories, they almost always end up having to.
+
+And again: this focuses on public, unauthenticated public APIs. I think we can agree this is not [GraphQL's Sweet Spot](https://productionreadygraphql.com/blog/2020-05-14-sweetspot).
+
+## Performance
+
+Many GraphQL critics talk about the N+1 problem and DataLoader. It's true that a GraphQL API must be ready for many query shapes and use cases. It is wise to implement efficient data loader for most fields through dataloaders. In fact, that's why I don't recommend using datafetching techniques that are overly coupled to the query shape, things like AST analysis, lookaheads, and things using context from sibling or parent fields.  But an honest observer should acknowledge that this is a problem with REST as well. As the complexity of an endpoint or operation increases, an engineer should expect to have to make more trade-offs in the name of performance.
+
+Overall I think dataloader is a requirement for GraphQL, and I agree with critics that it's part of the slight complexity add for a GraphQL API, even simple ones. Authorization N+1s are also an issue, though, again, REST has these issues as well.
+
+## Breaking Changes vs. Continuous Evolution
+
+One of the biggest "culture shocks" of adopting GraphQL is the concept of continuous evolution.  Because GraphQL has primitives for deprecating and aliasing fields, the notion of backwards compatibility becomes gray instead of black-and-white.  Want to add a field?  Great, do it.  Want to remove a field?  Add `@deprecated` and gradually stop querying them, eventually removing the field altogether.
+
+Arguably one of GraphQL's most powerful tool is around continuous evolution. The fact the client declaratively selects the API surface it is interested in means we can track with very high precision how our API is used. This allows us to actually **avoid** breaking changes, and make removals safely on fine grained parts of the schema.
+
+Breaking changes and deprecations suck. We all try to avoid them, and yes it's annoying for clients. But if anything GraphQL makes this easier, not harder.
```

Comment:
I would make that argument even stronger - in a case where you have a system, a service that runs for years and years.
there is no other technology other then GraphQL where you can know which fields its consumers are using in production right now.
Its hard to state the value of that knowledge on a legacy system